src/envoy/http/authn/http_filter.cc

```diff
@@ -48,49 +49,43 @@ FilterHeadersStatus AuthenticationFilter::decodeHeaders(HeaderMap& headers,
   filter_context_.reset(new Istio::AuthN::FilterContext(
       &headers, decoder_callbacks_->connection(), filter_config_));
 
-  authenticator_ = createPeerAuthenticator(
-      filter_context_.get(),
-      [this](bool success) { onPeerAuthenticationDone(success); });
-  authenticator_->run();
-
-  if (state_ == State::COMPLETE) {
-    return FilterHeadersStatus::Continue;
+  Payload payload;
+  authenticator_ = createPeerAuthenticator(filter_context_.get());
+  if (!authenticator_->run(&payload)) {
+    rejectRequest("Peer authentication failed.");
+    stopped_ = true;
+    return FilterHeadersStatus::StopIteration;
   }
 
-  stopped_ = true;
-  return FilterHeadersStatus::StopIteration;
-}
+  authenticator_ = createOriginAuthenticator(filter_context_.get());
```

Comment:
You can remove authenticator_, simply call
bool success = createOriginAuthenticator(filter_context_.get())->run(&payload);

Or even combine both into one, as we may not need different error rejection message.

bool success = createPeerAuthenticator(filter_context_.get())->run(&payload) 
&&createOriginAuthenticator(filter_context_.get())->run(&payload)